--- clipper.cpp.original	2016-12-04 14:15:18.000000000 -0800
+++ clipper.cpp	2016-12-18 23:30:12.626066121 -0800
@@ -2712,7 +2712,11 @@
 			e->Dx < horzEdge->NextInLML->Dx) break;
 
     if (horzEdge->OutIdx >= 0 && !IsOpen)  //note: may be done multiple times
-		{
+		{
+#ifdef use_xyz
+            if (dir == dLeftToRight) SetZ(e->Curr, *horzEdge, *e);
+            else SetZ(e->Curr, *e, *horzEdge);
+#endif
             op1 = AddOutPt(horzEdge, e->Curr);
 			TEdge* eNextHorz = m_SortedEdges;
 			while (eNextHorz)
@@ -3039,6 +3043,9 @@
       {
         e->Curr.X = TopX( *e, topY );
         e->Curr.Y = topY;
+#ifdef use_xyz
+        e->Curr.Z = topY == e->Top.Y ? e->Top.Z : (topY == e->Bot.Y ? e->Bot.Z : 0);
+#endif
       }
 
       //When StrictlySimple and 'e' is being touched by another edge, then
